### PR TITLE
Created test in log.go for merge commits

### PR DESCRIPTION
**Pull request by neilnaveen**
- Created test with a git tree containing a merge commit
- Added additional edge cases for regular git tree

### CODE REVIEW

**Comment by adityasaky** [comment]
cc @patzielinski

File: internal/gitinterface/log_test.go [hunk 1]
```diff
@@ -66,15 +66,21 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		allCommits = append(allCommits, commit)
+	allCommits, err := getCommitsFromCommitIds(commitIDs, repo)
+	if err != nil {
+		t.Fatal(err)
 	}
+	// Git tree structure with their commit trees and their values :
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:73]:
```suggestion
	// Git tree structure with their commit trees and their values:
```

File: internal/gitinterface/log_test.go [hunk 2]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
 		})
 		assert.Equal(t, expectedCommits, commits)
 	})
+	t.Run("Get commits from invalid range", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:111]:
```suggestion

	t.Run("Get commits from invalid range", func(t *testing.T) {
```

File: internal/gitinterface/log_test.go [hunk 3]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		commits, err := GetCommitsBetweenRange(repo, plumbing.ZeroHash, plumbing.ZeroHash)
+		assert.NotNil(t, err)
+		assert.Nil(t, commits)
+	})
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:115]:
```suggestion

	})
```

File: internal/gitinterface/log_test.go [hunk 4]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
 		})
 		assert.Equal(t, expectedCommits, commits)
 	})
+	t.Run("Get commits from invalid range", func(t *testing.T) {
+		commits, err := GetCommitsBetweenRange(repo, plumbing.ZeroHash, plumbing.ZeroHash)
+		assert.NotNil(t, err)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:113]: Can you `assert.ErrorIs` here?

File: internal/gitinterface/log_test.go [hunk 5]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		assert.Nil(t, commits)
+	})
+	t.Run("Get commits from non-existent commit", func(t *testing.T) {
+		nonExistentHash := plumbing.NewHash("1234567890123456789012345678901234567890")
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:117]: Let's use `EmptyBlob` or something here.

File: internal/gitinterface/log_test.go [hunk 6]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		commits, err := GetCommitsBetweenRange(repo, nonExistentHash, plumbing.ZeroHash)
+		assert.NotNil(t, err)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:119]: `assert.ErrorIs`?

File: internal/gitinterface/log_test.go [hunk 7]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	emptyBlobHash, err := WriteBlob(repo, []byte{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	treeHashes := make([]plumbing.Hash, 0, 6)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:149]: Can this be moved into a helper that accepts `n` for number of trees? Then that can be reused by the previous test too.

File: internal/gitinterface/log_test.go [hunk 8]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	})
+	t.Run("Get commits from non-existent commit", func(t *testing.T) {
+		nonExistentHash := plumbing.NewHash("1234567890123456789012345678901234567890")
+		commits, err := GetCommitsBetweenRange(repo, nonExistentHash, plumbing.ZeroHash)
+		assert.NotNil(t, err)
+		assert.Nil(t, commits)
+	})
+}
+
+func TestGetCommitsBetweenRangeForMergeCommits(t *testing.T) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:124]: I see why this is a separate test but a part of me wonders about keeping them organized by making this a `t.Run` in the previous one. :thinking:

File: internal/gitinterface/log_test.go [hunk 9]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	//    │              │  │              │
+	//    └─► commit 2 ◄─┘  └─► commit 3 ◄─┘
+	// 		      │              │
+	// 		 	  └─► commit 1 ◄─┘
+
+	allCommits, err := getCommitsFromCommitIds(commitIDs, repo)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Run("Get all commits from merge commit 5", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:231]:
```suggestion

	t.Run("Get all commits from merge commit 5", func(t *testing.T) {
```

File: internal/gitinterface/log_test.go [hunk 10]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	}
+
+	commitIDs = append(commitIDs, ref.Hash())
+
+	// Git tree with merge commit structure without its commit trees and its values :
+	//
+	//  commit 4       commit 5         commit 6
+	//    │              │  │              │
+	//    └─► commit 2 ◄─┘  └─► commit 3 ◄─┘
+	// 		      │              │
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:224]: I think the spacing on this diagram is off for commit 1 to commit 2

File: internal/gitinterface/log_test.go [hunk 11]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		childRefName := plumbing.ReferenceName(fmt.Sprintf("refs/heads/child%d", i))
+		childRef, err := applyCommitAndRefreshRef(repo, commit, plumbing.NewHashReference(childRefName, plumbing.ZeroHash))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		commitIDs = append(commitIDs, childRef.Hash())
+		children = append(children, childRef.Hash())
+	}
+	// creating a merge commit from the two children
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:189]:
```suggestion

	// creating a merge commit from the two children
```

File: internal/gitinterface/log_test.go [hunk 12]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	commit = CreateCommitObject(testGitConfig, treeHashes[5], []plumbing.Hash{children[1]}, fmt.Sprintf("Test commit %v", 6), testClock)
+
+	ref, err = applyCommitAndRefreshRef(repo, commit, ref)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	commitIDs = append(commitIDs, ref.Hash())
+
+	// Git tree with merge commit structure without its commit trees and its values :
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:219]:
```suggestion
	// Git tree with merge commit structure without its commit trees and its values:
```

File: internal/gitinterface/log_test.go [hunk 13]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		assert.Nil(t, err)
+		expectedCommits := []*object.Commit{allCommits[3], allCommits[2], allCommits[1], allCommits[0]}
+		sort.Slice(expectedCommits, func(i, j int) bool {
+			return expectedCommits[i].ID().String() < expectedCommits[j].ID().String()
+		})
+
+		assert.Equal(t, expectedCommits, commits)
+		assert.Nil(t, err)
+	})
+	t.Run("Get all commits from merge commit 6", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:242]:
```suggestion

	t.Run("Get all commits from merge commit 6", func(t *testing.T) {
```

File: internal/gitinterface/log_test.go [hunk 14]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		commits, err := GetCommitsBetweenRange(repo, commitIDs[3], plumbing.ZeroHash)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:243]: This invocation looks identical to the one for merge commit 5

Comment (adityasaky) [reply internal/gitinterface/log_test.go:243]: Also, is 6 a merge commit? In the diagram its parent seems to be commit 3 only?

File: internal/gitinterface/log_test.go [hunk 15]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		expectedCommits := []*object.Commit{allCommits[3], allCommits[2], allCommits[1], allCommits[0]}
+		sort.Slice(expectedCommits, func(i, j int) bool {
+			return expectedCommits[i].ID().String() < expectedCommits[j].ID().String()
+		})
+		assert.Equal(t, expectedCommits, commits)
+		assert.Nil(t, err)
+	})
+}
+
+func getCommitsFromCommitIds(commitIDs []plumbing.Hash, repo *git.Repository) ([]*object.Commit, error) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:254]: I suspect this could be a wrapper around `GetCommit` and actually exposed as part of gitinterface.

File: internal/gitinterface/log_test.go [hunk 16]
```diff
@@ -102,4 +108,171 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		if err != nil {
+			return nil, err
+		}
+
+		allCommits = append(allCommits, commit)
+	}
+	return allCommits, nil
+}
+
+func applyCommitAndRefreshRef(repo *git.Repository, commit *object.Commit, ref *plumbing.Reference) (*plumbing.Reference, error) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:267]: 1. Do we need to return a new plumbing.Reference? Since we're passing in the pointer of the ref, just updating it should do?
2. If this is only used in tests, let's pass in `t`, register this as a helper `t.Helper()`, and abort in case of errors `t.Fatal`. If 1 holds and with this change, we don't need to return anything at all here.

Comment (adityasaky) [reply internal/gitinterface/log_test.go:267]: Also, it'd be neat to track other places this helper could be used. I think there are certainly other tests that could reuse this.

Comment (adityasaky) [reply internal/gitinterface/log_test.go:267]: They don't need to be updated in this PR though.

**Review by adityasaky — CHANGES_REQUESTED** [review]
Thanks @neilnaveen! Some questions around the new tests and some formatting suggestions.

File: internal/gitinterface/log_test.go (hunk 10 above)
Comment (adityasaky) [reply internal/gitinterface/log_test.go:224]: It's off in https://github.com/gittuf/gittuf/blob/94b5e2a50f068974d37bdd0a6e0ed8c03206e158/internal/gitinterface/log_test.go too though :\

The issue is there's a mix of `\t` and spaces. We can't control tab width, so I suggest using spaces throughout for these diagrams.

File: internal/gitinterface/log_test.go (hunk 14 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:243]: Yes, this is the wrong test name, I was thinking about something on the line of, "Get commits from leaf commit 6 to root commit 1"

File: internal/gitinterface/log_test.go (hunk 16 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Fixed

File: internal/gitinterface/log_test.go (hunk 15 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: I have made this an exported func

File: internal/gitinterface/log_test.go (hunk 14 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Fixed, changed the test name, and fixed the test to actually reflect test name. Merge commit 5's test has also been fixed, before it was not actually testing the merge commit, but something else.

File: internal/gitinterface/log_test.go (hunk 13 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 12 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 7 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 9 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:195]: fixed

File: internal/gitinterface/log_test.go (hunk 10 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 11 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:161]: fixed

File: internal/gitinterface/log_test.go (hunk 5 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 3 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:106]: fixed

File: internal/gitinterface/log_test.go (hunk 2 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:102]: fixed

File: internal/gitinterface/log_test.go (hunk 1 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go (hunk 4 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: fixed

File: internal/gitinterface/log_test.go [hunk 17]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		assert.Nil(t, err)
+		expectedCommits := []*object.Commit{allCommits[5], allCommits[2], allCommits[0]}
+		sort.Slice(expectedCommits, func(i, j int) bool {
+			return expectedCommits[i].ID().String() < expectedCommits[j].ID().String()
+		})
+		assert.Equal(t, expectedCommits, commits)
+	})
+}
+
+func GetCommitsFromCommitIds(commitIDs []plumbing.Hash, repo *git.Repository) ([]*object.Commit, error) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:217]:
```suggestion
func GetCommitsFromCommitIDs(commitIDs []plumbing.Hash, repo *git.Repository) ([]*object.Commit, error) {
```

Couple other thoughts with this one:
1. For consistency, please reorder the arguments so we pass in the repo first. This aligns with other APIs in this package.
2. We shouldn't have an exported function in a `_test` file. This should go in `commit.go`.
3. We really ought to be testing this function.

File: internal/gitinterface/log_test.go (hunk 8 above)
Comment (adityasaky) [reply internal/gitinterface/log_test.go:115]: I think I've come around on this one given the difference in setup.

File: internal/gitinterface/log_test.go [hunk 18]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	applyCommitAndRefreshRef(t, repo, commit, ref)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	commitIDs = append(commitIDs, ref.Hash())
+
+	// creating two children from the first commit
+
+	childrenCommitIds, children := createChildrenCommits(t, repo, ref, treeHashes, 2)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:157]:
```suggestion
	childrenCommitIDs, children := createChildrenCommits(t, repo, ref, treeHashes, 2)
```

File: internal/gitinterface/log_test.go [hunk 19]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	treeHashes, err := createTreeHashes(repo, emptyBlobHash, 6)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// creating the first commit
+
+	commit := CreateCommitObject(testGitConfig, treeHashes[0], []plumbing.Hash{ref.Hash()}, fmt.Sprintf("Test commit %v", 1), testClock)
+
+	applyCommitAndRefreshRef(t, repo, commit, ref)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:148]: Strictly speaking, we only want the commits, no? We don't need to actually apply them to a ref? The commits can just live in the object store.

File: internal/gitinterface/log_test.go (hunk 18 above)
Comment (adityasaky) [reply internal/gitinterface/log_test.go:157]: Also, for consistency and readability, please update `children` (the second return value) to indicate what it is. I'm confused because it seems to be identical to `childrenCommitIDs`?

File: internal/gitinterface/log_test.go [hunk 20]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	// creating a merge commit from the two children
+
+	commit = CreateCommitObject(testGitConfig, treeHashes[4], children, fmt.Sprintf("Test commit %v", 5), testClock)
+
+	applyCommitAndRefreshRef(t, repo, commit, ref)
+	commitIDs = append(commitIDs, ref.Hash())
+
+	// creating children for each of the two children
+
+	commit = CreateCommitObject(testGitConfig, treeHashes[3], []plumbing.Hash{children[0]}, fmt.Sprintf("Test commit %v", 4), testClock)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:170]: Could you indicate in a comment what each index in `children` maps to in the diagram? It's hard to keep track atm.

File: internal/gitinterface/log_test.go [hunk 21]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	//            │              │
+	//            └─► commit 1 ◄─┘
+
+	allCommits, err := GetCommitsFromCommitIds(commitIDs, repo)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("Get all commits from merge commit 5", func(t *testing.T) {
+		commits, err := GetCommitsBetweenRange(repo, commitIDs[4], plumbing.ZeroHash)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:196]: Here as well, a comment to indicate what each index maps to would be helpful to know what commits are returned in the diagram. That should give us a good idea of whether the behavior aligns with our expectations.

File: internal/gitinterface/log_test.go [hunk 22]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	allCommits := make([]*object.Commit, 0, len(commitIDs))
+	for _, commitID := range commitIDs {
+		commit, err := GetCommit(repo, commitID)
+		if err != nil {
+			return nil, err
+		}
+
+		allCommits = append(allCommits, commit)
+	}
+	return allCommits, nil
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:227]:
```suggestion

	return allCommits, nil
```

File: internal/gitinterface/log_test.go [hunk 23]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+}
+
+func applyCommitAndRefreshRef(t *testing.T, repo *git.Repository, commit *object.Commit, ref *plumbing.Reference) {
+	t.Helper()
+	if _, err := ApplyCommit(repo, commit, ref); err != nil {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:232]:
```suggestion

	if _, err := ApplyCommit(repo, commit, ref); err != nil {
```

File: internal/gitinterface/log_test.go [hunk 24]
```diff
@@ -102,4 +98,182 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+
+	for i := 1; i <= numChildren; i++ {
+		commit := CreateCommitObject(testGitConfig, treeHashes[i], []plumbing.Hash{ref.Hash()}, fmt.Sprintf("Test commit %v", i+1), testClock)
+		childRefName := plumbing.ReferenceName(fmt.Sprintf("refs/heads/child%d", i))
+		childRef := plumbing.NewHashReference(childRefName, commit.Hash)
+
+		applyCommitAndRefreshRef(t, repo, commit, childRef)
+
+		commitIDs = append(commitIDs, childRef.Hash())
+		children = append(children, childRef.Hash())
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:276]: Are you returning this twice to separate handling children at during invocation? I wonder if some things could be moved around there instead?

**Review by adityasaky — CHANGES_REQUESTED** [review]
Some more comments, I think this is getting a lot closer!

File: internal/gitinterface/log_test.go (hunk 20 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: I did this, and just to clear things up, I made a test for every single commit, so people can understand what connects to what, and how it works.

File: internal/gitinterface/log_test.go (hunk 21 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Done

File: internal/gitinterface/log_test.go (hunk 22 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:259]: Fixed

File: internal/gitinterface/log_test.go (hunk 23 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:265]: Fixed

File: internal/gitinterface/log_test.go (hunk 24 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: I had refactored the creation of children into the function, and I overlooked that, the result is now just one slice of hashes that is be appended to the commitID's slice in the main test.

File: internal/gitinterface/log_test.go (hunk 18 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Fixed

File: internal/gitinterface/log_test.go (hunk 17 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: would you like me to move this function to commit.go and in a later PR test it, or should I just keep it as an non-exported function for now?

File: internal/gitinterface/log_test.go [hunk 25]
```diff
@@ -36,19 +36,9 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
-			objects = append(objects, object.TreeEntry{Name: fmt.Sprintf("%d", j+1), Hash: emptyBlobHash})
-		}
-
-		treeHash, err := WriteTree(repo, objects)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		treeHashes = append(treeHashes, treeHash)
+	treeHashes, err := createTreeHashes(repo, emptyBlobHash, 5)
```
[the remark this answers is not in the document]
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Fixed

File: internal/gitinterface/log_test.go (hunk 19 above)
Comment (adityasaky) [reply internal/gitinterface/log_test.go:148]: bump

**Comment by neilnaveen** [comment]
The test are failing, because when I am calling GetCommitsBetweenRange from a given commit. I am getting it with the second parameter as a zero hash, I am not getting the first commit in the commit tree, which is expected. Am I creating the first commit incorrectly? @adityasaky

File: internal/gitinterface/log_test.go [hunk 26]
```diff
@@ -102,4 +99,218 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	commitIDs = append(commitIDs, commitHash)
+
+	// creating two children from the first commit
+
+	// in the visual, these will be commit 2 and commit 3
+
+	children := createChildrenCommits(t, repo, ref, treeHashes, 2)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:159]: `ref` is never updated here, so commit 2 and 3 are created without a parent, right?

Comment (adityasaky) [reply internal/gitinterface/log_test.go:159]: I suggest ripping `ref` out entirely, we only need the commits themselves.

Comment (neilnaveen) [reply internal/gitinterface/log_test.go:159]: I forgot about that, will change

Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Fixed

File: internal/gitinterface/log_test.go [hunk 27]
```diff
@@ -102,4 +99,218 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	return treeHashes
+}
+
+func createChildrenCommits(t *testing.T, repo *git.Repository, treeHashes []plumbing.Hash, numChildren int) []plumbing.Hash {
+	t.Helper()
+
+	children := make([]plumbing.Hash, 0, numChildren)
+
+	for i := 1; i <= numChildren; i++ {
+		commit := CreateCommitObject(testGitConfig, treeHashes[i], []plumbing.Hash{}, fmt.Sprintf("Test commit %v", i+1), testClock)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:307]: this also ends up with no parents, you want to pass in the parent commit ID(s) here.

Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: @adityasaky I have done this, is this the change you mean?

File: internal/gitinterface/log_test.go [hunk 28]
```diff
@@ -102,4 +99,218 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+			t.Errorf("Expected and actual commits do not match. Expected: %v, Got: %v", expectedCommits, commits)
+		}
+	})
+
+	t.Run("Test commit 3", func(t *testing.T) {
+		commits, err := GetCommitsBetweenRange(repo, commitIDs[2], plumbing.ZeroHash)
+		assert.Nil(t, err)
+		// commit 3 is the second child of commit 1, so only it and commit 1 should be returned
+		expectedCommits := []*object.Commit{allCommits[0], allCommits[2]}
+		if !reflect.DeepEqual(expectedCommits, commits) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:232]: any specific reason for not using `assert.Equal` here? I think this was originally using assert?

File: internal/gitinterface/log_test.go [hunk 29]
```diff
@@ -102,4 +99,218 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+	//    │              │  │              │
+	//    └─► commit 2 ◄─┘  └─► commit 3 ◄─┘
+	//            │              │
+	//            └─► commit 1 ◄─┘
+
+	allCommits, err := GetCommitsFromCommitIDs(commitIDs, repo)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Run("Test commit 1", func(t *testing.T) {
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:207]:
```suggestion

	t.Run("Test commit 1", func(t *testing.T) {
```

as a rule of thumb, please insert a newline before every new "block" like a new `t.Run` or whatever.

**Review by adityasaky — COMMENTED** [review]
final comments I think!

File: internal/gitinterface/log_test.go [hunk 30]
```diff
@@ -102,4 +99,218 @@ func TestGetCommitsBetweenRange(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	commitIDs = append(commitIDs, commitHash)
+
+	// creating two children from the first commit
+
+	// in the visual, these will be commit 2 and commit 3
+
+	children := createChildrenCommits(t, repo, treeHashes, commit.Hash, 2)
```
Comment (adityasaky) [inline internal/gitinterface/log_test.go:159]: Hmm, I'm guessing here but I think you want to use `commitHash` here instead of `.Hash`. This is because `commit` is not updated after `WriteCommit` so it has no notion of its hash in the storer.

File: internal/gitinterface/log_test.go (hunk 28 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:232]: Assert errors out for this test, at least in my editor because it cannot compare expected commits with the returned commits, since the commits have unexported fields, which assert cannot process.

File: internal/gitinterface/log_test.go (hunk 29 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go:208]: Fixed

File: internal/gitinterface/log_test.go (hunk 30 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Done

**Comment by neilnaveen** [comment]
> The test are failing, because when I am calling GetCommitsBetweenRange from a given commit. I am getting it with the second parameter as a zero hash, I am not getting the first commit in the commit tree, which is expected. Am I creating the first commit incorrectly? @adityasaky

I am guessing that they were just flakey test, because now they are running properly 😆

File: internal/gitinterface/log_test.go (hunk 19 above)
Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: Taken out

File: internal/gitinterface/log_test.go (hunk 28 above)
Comment (adityasaky) [reply internal/gitinterface/log_test.go:233]: Can you paste the error here? We don't need to assert private fields, and we use assert with similar inputs in other tests.

Comment (neilnaveen) [reply internal/gitinterface/log_test.go]: @adityasaky I just ran it again, with assert, and it worked. I am guessing my test were flakey. Changed the reflect.DeepEqual to assert.Equal

**Comment by adityasaky** [comment]
Can you squash your commits? :)

**Comment by neilnaveen** [comment]
> Can you squash your commits? :)

Squashed
